File: src/main/java/com/dv/telegram/WikiBot.java
package com.dv.telegram;

import com.dv.telegram.command.BotSpecialCommands;
import com.dv.telegram.command.SpecialCommandResponse;
import com.dv.telegram.config.BotSettings;
import com.dv.telegram.data.*;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.lang3.StringUtils;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage;
import org.telegram.telegrambots.meta.api.objects.Message;
import org.telegram.telegrambots.meta.api.objects.Update;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.text.MessageFormat;
import java.util.List;
import java.util.Locale;
import java.util.Optional;
import java.util.regex.Pattern;

@Log4j2
public class WikiBot extends TelegramLongPollingBot {

    private static final String START_COMMAND = "/start";

    private final WikiBotConfig config;
    private WikiPagesDataList pages;
    private List<CityChatData> cityChats;
    private List<CountryChatData> countryChats;
    private WikiBotCommandsDataList commands;

    private final String botName;
    private final String botNameLowerCase;
    private final Pattern botNameWordPattern;
    private final String environmentName;

    private final BotSpecialCommands specialCommands;

    private final BotSettings settings; // settings cache in memory

    private final BotStatistics statistics;

    public WikiBot(
        WikiBotConfig config,
        GoogleSheetBotData botData
    ) {
        this(
            config,
            botData.getPages(),
            botData.getCityChats(),
            botData.getCountryChats(),
            botData.getCommands()
        );
    }

    public WikiBot(
        WikiBotConfig config,
        List<WikiPageData> wikiPagesData,
        List<CityChatData> cityChatsData,
        List<CountryChatData> countryChatsData,
        List<WikiBotCommandData> commands
    ) {
        super();

        this.config = config;
        this.pages = new WikiPagesDataList(wikiPagesData);
        this.cityChats = cityChatsData;
        this.countryChats = countryChatsData;
        this.commands = new WikiBotCommandsDataList(commands);

        this.botName = config.getBotName();
        this.botNameLowerCase = config.getBotName().toLowerCase(Locale.ROOT);
        this.botNameWordPattern = getBotNameFullWordPattern(botName);

        this.environmentName = config.getEnvironmentName();

        this.specialCommands = BotSpecialCommands.create(config);
        this.settings = BotSettings.create(config);

        this.statistics = new BotStatistics();
    }

    public String getBotToken() {
        return config.getBotToken();
    }

    public String getBotUsername() {
        return "dv_wiki_bot"; // todo: read from config if needed. Seems to be overridden by what is set by "/setname" in the @BotFather
    }

    public String getBotName() {
        return botName;
    }

    public String getEnvironmentName() {
        return environmentName;
    }

    public String getNotionToken() {
        return config.notionToken;
    }

    public String getNotionCityPageId() {
        return config.cityChatsPageId;
    }

    public String getNotionCityChatsToggleHeading1Text() {
        return config.cityChatsToggleHeading1Text;
    }

    public BotSpecialCommands getSpecialCommands() {
        return specialCommands;
    }

    public WikiBotCommandsDataList getCommands() {
        return commands;
    }

    public BotSettings getSettings() {
        return settings;
    }

    public BotStatistics getStatistics() {
        return statistics;
    }

    public static Pattern getBotNameFullWordPattern(String botName) {
        String botNameRegex = String.format("(?i).*\\b%s\\b.*", botName);
        int botNamePatternFlags = Pattern.CASE_INSENSITIVE | Pattern.UNICODE_CASE | Pattern.DOTALL;
        return Pattern.compile(botNameRegex, botNamePatternFlags); // see https://stackoverflow.com/a/43738714/8534088
    }

    public void onUpdateReceived(Update update) {
        // todo: env option to disable the bot globally!

        Message updateMessage = update.getMessage();
        if (!update.hasMessage() || !updateMessage.hasText()) {
            return;
        }

        String text = updateMessage.getText();
        String userName = updateMessage.getFrom().getUserName();

        MessageProcessingResult processingResult = getResponseText(text, userName);
        statistics.update(text, processingResult);

        if (!processingResult.messageIsForTheBot) { // message is not for the bot -> do nothing
            return;
        }

        // if the message is a reply, is for the bot and ReplyWhenNoAnswer is false, we still have to delete it
        Message replyToMessage = updateMessage.getReplyToMessage();
        boolean updateMessageIsReply = (replyToMessage != null);

        boolean deleteBotCallMessage = settings.getDeleteBotCallMessageOnMessageReply() && updateMessageIsReply;
        if (deleteBotCallMessage) {
            deleteBotCallMessage(updateMessage);
        }

        if (processingResult.hasNoResponse()) { // no response -> nothing to return
            return;
        }

        // command is for the bot and has the response -> send the response message
        try {
            SendMessage sendMessage = createSendMessage(updateMessage, replyToMessage, deleteBotCallMessage, processingResult);
            execute(sendMessage); // Call method to send the message
        }
        catch (TelegramApiException e) {
            log.error("TelegramApiException occurred on sending the bot response message", e); // todo: add which message has failed the bot
            throw new RuntimeException(e);
        }
    }

    private SendMessage createSendMessage(
        Message updateMessage,
        Message replyToMessage,
        boolean deleteBotCallMessage,
        MessageProcessingResult processingResult
    ) {
        String chatId = updateMessage.getChatId().toString();
        String responseText = processingResult.getResponseOrFail();

        Integer replyToMessageId = deleteBotCallMessage
            ? replyToMessage.getMessageId() // reply to the original message
            : updateMessage.getMessageId() // reply to the "call bot" message
        ;

        SendMessage sendMessage = new SendMessage();
        sendMessage.setChatId(chatId);
        sendMessage.setReplyToMessageId(replyToMessageId);
        sendMessage.setText(responseText);
        sendMessage.disableWebPagePreview();

        if (processingResult.useMarkdown) {
            sendMessage.setParseMode("Markdown");
        }

        return sendMessage;
    }

    private void deleteBotCallMessage(Message updateMessage) {
        DeleteMessage deleteMessage = new DeleteMessage(
            updateMessage.getChatId().toString(),
            updateMessage.getMessageId()
        );

        try {
            execute(deleteMessage);
        }
        catch (TelegramApiException e) {
            log.error("Failed to delete message", e);
        }
    }

    private boolean messageIsForTheBot(String lowerText) {
        if (lowerText.equals(START_COMMAND)) { // special case: /start command without bot name
            return true;
        }

        return switch (settings.getTriggerMode()) { // check whether the message contains the bot name
            case ANY_SUBSTRING -> lowerText.contains(botNameLowerCase);
            case STRING_START -> lowerText.startsWith(botNameLowerCase);
            case FULL_WORD -> botNameWordPattern.matcher(lowerText).matches();
        };
    }

    MessageProcessingResult getResponseText(String text, String userName) { // non-private for testing
        if (StringUtils.isBlank(text)) {
            return MessageProcessingResult.notForTheBot();
        }

        String lowerText = text.toLowerCase(Locale.ROOT);

        if (!messageIsForTheBot(lowerText)) { // only work when bot is mentioned by name
            return MessageProcessingResult.notForTheBot();
        }

        // special commands - not configured in the Google Sheet
        SpecialCommandResponse specialCommandResponse = specialCommands.getResponse(text, userName, this);
        if (specialCommandResponse.hasResponse()) { // special command received -> return response for the special command
            return MessageProcessingResult.specialCommand(specialCommandResponse.response, specialCommandResponse.useMarkdownInResponse);
        }

        // todo: think on executing commands.getResponseText and checking whether it is empty
        // normal commands - configured in the Google Sheet
        List<WikiBotCommandData> matchingCommands = commands.findMatches(lowerText);
        if (!matchingCommands.isEmpty()) { // matching command found -> only handle the command
            String commandAnswerText = commands.getResponseText(matchingCommands);

            return MessageProcessingResult.answerFound(commandAnswerText);
        }

        // wiki pages - configured in the Google Sheet
        Optional<String> wikiPageAnswerText = Optional.ofNullable(
            pages.getResponseText(lowerText)
        );

        // city chats - configured in the Google Sheet
        List<CityChatData> matchingCityChats = findMatchingCityChats(lowerText);
        Optional<String> cityChatsAnswerText = getCityChatsAnswerText(matchingCityChats);

        // country chats - configured in the Google Sheet
        List<CountryChatData> matchingCountryChats = findMatchingCountryChats(lowerText);
        Optional<String> countryChatsAnswerText = getCountryChatsAnswerText(matchingCountryChats);

        List<Optional<String>> answers = List.of(
            wikiPageAnswerText,
            cityChatsAnswerText,
            countryChatsAnswerText
        );

        return getResponseText(text, answers);
    }

    private MessageProcessingResult getResponseText(String text, List<Optional<String>> answerOptionals) {
        List<String> answers = answerOptionals
            .stream()
            .filter(Optional::isPresent)
            .map(Optional::get)
            .toList();

        if (answers.isEmpty()) { // no answers found
            Optional<String> noResultResponse = getNoResultResponse(text);
            return MessageProcessingResult.answerNotFound(noResultResponse);
        }

        String combinedAnswers = StringUtils.join(answers, "\n\n");
        return MessageProcessingResult.answerFound(combinedAnswers);
    }

    public GoogleSheetBotData loadBotDataFromGoogleSheet() { // does NOT reload the bot data itself
        return GoogleSheetLoader.readGoogleSheet(config);
    }

    public boolean reloadBotDataFromGoogleSheet() {
        try {
            GoogleSheetBotData botData = loadBotDataFromGoogleSheet();
            this.pages = new WikiPagesDataList(botData.getPages());
            this.cityChats = botData.getCityChats();
            this.countryChats = botData.getCountryChats();
            this.commands = new WikiBotCommandsDataList(botData.getCommands());

            return true;
        }
        catch (Exception e) {
            return false;
        }
    }

    private List<CityChatData> findMatchingCityChats(String text) {
        return cityChats
            .stream()
            .filter(cityChat -> cityChat.isPresentIn(text))
            .toList();
    }

    private List<CountryChatData> findMatchingCountryChats(String text) {
        return countryChats
            .stream()
            .filter(countryChat -> countryChat.isPresentIn(text))
            .toList();
    }

    private Optional<String> getCityChatsAnswerText(List<CityChatData> matchingCityChats) {
        if (matchingCityChats.isEmpty()) {
            return Optional.empty();
        }

        List<String> multilineAnswers = matchingCityChats
            .stream()
            .map(CityChatData::getChatsAnswer)
            .toList();

        String answer = StringUtils.join(multilineAnswers, "\n\n");
        return Optional.of(answer);
    }

    private Optional<String> getCountryChatsAnswerText(List<CountryChatData> matchingCityChats) {
        if (matchingCityChats.isEmpty()) {
            return Optional.empty();
        }

        List<String> multilineAnswers = matchingCityChats
            .stream()
            .map(CountryChatData::getChatsAnswer)
            .toList();

        String answer = StringUtils.join(multilineAnswers, "\n\n");
        return Optional.of(answer);
    }

    private Optional<String> getNoResultResponse(String text) {
        log.info("Unknown command for the bot: {}", text);

        if (settings.getReplyWhenNoAnswer()) { // reply on no answer
            String noResultAnswer = getNoResultAnswer(text);
            return Optional.of(noResultAnswer);
        }

        // no reply on no answer
        return Optional.empty();
    }

    String getNoResultAnswer(String text) { // not private for testing
        return MessageFormat.format(settings.getNoAnswerReply(), botName, text);
    }
}

File: src/main/kotlin/com/dv/telegram/WikiBot.kt
package com.dv.telegram

import com.dv.telegram.command.BotSpecialCommands
import com.dv.telegram.config.BotSettings
import com.dv.telegram.data.BotAnswerData
import com.dv.telegram.statistics.BotStatistics
import com.dv.telegram.tabs.BotAnswerTabData
import com.dv.telegram.tabs.TabConfigs
import com.dv.telegram.tabs.TabData
import org.apache.logging.log4j.kotlin.Logging
import org.telegram.telegrambots.bots.TelegramLongPollingBot
import org.telegram.telegrambots.meta.api.methods.GetMe
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.exceptions.TelegramApiException

@Suppress("TooManyFunctions")
class WikiBot(
    val context: WikiBotsContext,
    private val config: WikiBotConfig,
    commandTabsData: List<TabData>,
    dataTabsData: List<TabData>
) : TelegramLongPollingBot(), Logging {
    var commandTabs: List<BotAnswerTabData<BotAnswerData>> // can be reloaded from Google Sheet
        private set

    var dataTabs: List<BotAnswerTabData<BotAnswerData>> // can be reloaded from Google Sheet
        private set

    val botName: String

    val environmentName: String

    val specialCommands: BotSpecialCommands

    val settings: BotSettings // settings cache in memory

    val statistics: BotStatistics

    val messageProcessor: WikiBotMessageProcessor

    var lastMessageLog: MessageProcessingLog? = null

    var telegramName: String? = null

    constructor(
        context: WikiBotsContext,
        config: WikiBotConfig,
        botTabsData: GoogleSheetBotTabsData
    ) : this(
        context,
        config,
        botTabsData.commandTabs,
        botTabsData.dataTabs
    )

    init {
        context.addBot(this)

        commandTabs = BotAnswerTabData.fromTabDataList(commandTabsData)
        dataTabs = BotAnswerTabData.fromTabDataList(dataTabsData)

        botName = config.botName

        environmentName = config.environmentName

        specialCommands = BotSpecialCommands.create(config)
        settings = BotSettings.create(config)

        statistics = BotStatistics()

        messageProcessor = WikiBotMessageProcessor(this)
    }

    // additional getters
    val tabConfigs: TabConfigs
        get() = config.sheets

    val notionToken: String
        get() = config.notionToken

    val notionCityPageId: String
        get() = config.cityChatsPageId

    val notionCityChatsToggleHeading1Text: String
        get() = config.cityChatsToggleHeading1Text

    // functions overridden from Telegram API (TelegramLongPollingBot and its parent hierarchy)
    override fun getBotToken(): String {
        return config.botToken
    }

    override fun getBotUsername(): String {
        return "dv_wiki_bot" // todo: read from config if needed. Seems to be overridden by what is set by "/setname" in the @BotFather
    }

    override fun onUpdateReceived(update: Update) {
        // todo: env option to disable the bot globally!
        val updateMessage = update.message
        if (!update.hasMessage() || !updateMessage.hasText()) {
            return
        }

        val text = updateMessage.text
        val userName = updateMessage.from.userName ?: "" // userName is nullable, user can have no username in Telegram

        val processingResult = processMessage(text, userName, update)

        if (!processingResult.messageIsForTheBot) { // message is not for the bot -> do nothing
            return
        }

        // if the message is a reply, is for the bot and ReplyWhenNoAnswer is false, we still have to delete it
        val replyToMessage = updateMessage.replyToMessage
        val updateMessageIsReply = (replyToMessage != null)

        val deleteBotCallMessage = settings.deleteBotCallMessageOnMessageReply && updateMessageIsReply
        if (deleteBotCallMessage) {
            deleteBotCallMessage(updateMessage)
        }

        if (processingResult.hasNoResponse()) { // no response -> nothing to return
            return
        }

        // command is for the bot and has the response -> send the response message
        try {
            val sendMessages = createSendMessage(updateMessage, replyToMessage, deleteBotCallMessage, processingResult)

            for (sendMessage in sendMessages) { // messages are split into 4096 characters
                execute(sendMessage) // Call method to send the message
            }
        }
        catch (e: TelegramApiException) {
            logger.error("TelegramApiException occurred on sending the bot response message", e) // todo: add which message has failed the bot
            throw RuntimeException(e)
        }
    }

    private fun createSendMessage(
        updateMessage: Message,
        replyToMessage: Message?,
        deleteBotCallMessage: Boolean,
        processingResult: MessageProcessingResult
    ): List<SendMessage> {
        val chatId = updateMessage.chatId.toString()
        val responseText = processingResult.getResponseOrFail()

        // todo: probably some more intellectual split like on new lines near the limits
        val responseTexts = responseText.chunked(MAX_CHARACTERS_IN_MESSAGE)

        val messages = mutableListOf<SendMessage>()

        for (messageText in responseTexts) {
            val replyToMessageId =
                if (deleteBotCallMessage) { // reply to the original message
                    replyToMessage!!.messageId
                }
                else { // reply to the "call bot" message
                    updateMessage.messageId
                }

            val sendMessage = SendMessage()
            sendMessage.chatId = chatId
            sendMessage.replyToMessageId = replyToMessageId
            sendMessage.text = messageText
            sendMessage.disableWebPagePreview()

            // todo: markdown can be broken if it is split into different message parts. Probably turn it off if responseTexts.size > 1
            if (processingResult.useMarkdown) {
                sendMessage.parseMode = "Markdown"
            }

            messages.add(sendMessage)
        }

        return messages
    }

    private fun deleteBotCallMessage(updateMessage: Message) {
        val deleteMessage = DeleteMessage(
            updateMessage.chatId.toString(),
            updateMessage.messageId
        )

        try {
            execute(deleteMessage)
        }
        catch (e: TelegramApiException) {
            logger.error("Failed to delete message", e)
        }
    }

    fun processMessage(text: String, userName: String, update: Update): MessageProcessingResult { // non-private for testing
        val result = messageProcessor.processMessage(text, userName, update)

        statistics.update(text, result) // moved into this method to be updated in the test

        if (result.messageIsForTheBot) {
            lastMessageLog = MessageProcessingLog.create(this, update, result)
        }

        return result
    }

    fun getTelegramUserName(): String {
        if (telegramName != null) {
            return telegramName!!
        }

        return try {
            val getMe = GetMe()
            val getMeResult = execute(getMe)

            telegramName = getMeResult.userName

            telegramName!!
        }
        catch (e: Exception) {
            logger.error("Error when getting Telegram user name for bot $botName", e)
            "unknown"
        }
    }

    fun getCommandTabNames(): List<String> = config.sheets.commandTabs.map { it.tabName }

    fun reloadBotDataFromGoogleSheet(): Boolean {
        return try {
            val botData = loadBotDataFromGoogleSheet()

            commandTabs = BotAnswerTabData.fromTabDataList(botData.commandTabs)
            dataTabs = BotAnswerTabData.fromTabDataList(botData.dataTabs)

            true
        }
        catch (e: Exception) {
            false
        }
    }

    fun loadBotDataFromGoogleSheet(): GoogleSheetBotTabsData { // does NOT reload the bot data itself
        return GoogleSheetLoader.readGoogleSheetTabs(config)
    }

    companion object {
        private const val MAX_CHARACTERS_IN_MESSAGE = 4096
    }
}
